source/rofi-icon-fetcher.c

```diff
@@ -91,6 +98,152 @@ typedef struct {
+  
+  if (!dir) {
+    g_free(thumb_path);
+    return;
+  }
+
+  const gchar *dirent;
+  
+  while ((dirent = g_dir_read_name(dir))) {
+    if (!g_str_has_suffix(dirent, THUMBNAILER_EXTENSION))
+      continue;
+
+    gchar *filename = g_build_filename(thumb_path, dirent, NULL);
+    GKeyFile *key_file = g_key_file_new();
+    GError *error = NULL;
+
+    if (!g_key_file_load_from_file(key_file, filename, 0, &error)) {
+      g_warning("Error loading thumbnailer %s: %s", filename, error->message);
+      g_error_free(error);
+    } else {
+      gchar *command = g_key_file_get_string(
+        key_file, THUMBNAILER_ENTRY_GROUP, "Exec", NULL);
+      gchar **mime_types = g_key_file_get_string_list(
+        key_file, THUMBNAILER_ENTRY_GROUP, "MimeType", NULL, NULL);
+
+      if (mime_types && command) {
+        guint i;
+        for (i = 0; mime_types[i] != NULL; i++) {
+          if (!g_hash_table_lookup(
+              rofi_icon_fetcher_data->thumbnailers, mime_types[i])) {
+            g_info("Loading thumbnailer %s for mimetype %s", filename, mime_types[i]);
+            g_hash_table_insert(
+              rofi_icon_fetcher_data->thumbnailers,
+              g_strdup(mime_types[i]),
+              g_strdup(command)
+            );
+          }
+        }
+      }
+
+      if (mime_types) g_strfreev(mime_types);
+      if (command) g_free(command);
+    }
+
+    g_key_file_free(key_file);
+    g_free(filename);
+  }
+
+  g_dir_close(dir);
+  g_free(thumb_path);
+}
+
+static gchar** setup_thumbnailer_command(const gchar *command,
+                                         const gchar *filename,
+                                         const gchar *encoded_uri,
+                                         const gchar *output_path,
+                                         int size) {
+  gchar **command_parts = g_strsplit(command, " ", 0);
+  guint command_parts_count = g_strv_length(command_parts);
+  
+  gchar **command_args = NULL; 
+  
+  if (command_parts) {
+    command_args = malloc(sizeof(gchar*) * (command_parts_count + 3 + 1));
+    
+    // set process niceness value to 19 (low priority)
+    guint current_index = 0;
+    
+    command_args[current_index++] = strdup("nice");
+    command_args[current_index++] = strdup("-n");
+    command_args[current_index++] = strdup("19");
+    
+    // add executable and arguments of the thumbnailer to the list
+    guint i;
+    for (i = 0; command_parts[i] != NULL; i++) {
+      if (strcmp(command_parts[i], "%i") == 0) {
+        command_args[current_index++] = strdup(filename);
+      } else if (strcmp(command_parts[i], "%u") == 0) {
+        command_args[current_index++] = strdup(encoded_uri);
+      } else if (strcmp(command_parts[i], "%o") == 0) {
+        command_args[current_index++] = strdup(output_path);
+      } else if (strcmp(command_parts[i], "%s") == 0) {
+        char size_str[33];
+        snprintf(size_str, 33, "%d", size);
+        command_args[current_index++] = strdup(size_str);
```

Comment:
`g_strdup_printf` should be more idiomatic here.

Maybe also `g_malloc0` and `g_strdup` in the other calls for consistency?